### ISSUE TITLE
Merino integration test caching::redis_tests::bad_cache_data_is_handled is flaky

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-services/merino#183]
The test `caching::redis_tests::bad_cache_data_is_handled` in the `merino_integration_tests` crate fails for me about 15% of the time. We should figure out why and get it to pass more reliably.

The test expects an error to be reported, but I see it fail sometimes with Merino incorrectly returning valid results. I suspect this is a race condition, but I haven't been able to track it down yet. Notably, it doesn't happen when I run the test in isolation, so it may be an issue with another test running in parallel.

Note: I suspect this is an issue in the tests, not an underlying problem in the code.

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/CONSVC-1325)

**Comment by data-sync-user** [comment]
➤ Michael Cooper commented:

I have learned something about why this test is failing. The test makes a request to put an item in the cache, gets the cache key by fetching the first key from Redis, and then modifies the value it finds there.

After we introduced Redis cache locking, sometimes the key it finds is the key for the lock instead of the key for the cached value. This is not what the test expects, so it fails.
